Release/src/http/client/winhttppal/winhttppal.cpp

```diff
@@ -0,0 +1,3804 @@
+
+    case CURLINFO_HEADER_IN:
+        text = "=> Receive header";
+        break;
+    case CURLINFO_HEADER_OUT:
+        text = "=> Send header";
+        break;
+    case CURLINFO_DATA_OUT:
+        text = "=> Send data";
+        break;
+    case CURLINFO_SSL_DATA_OUT:
+        text = "=> Send SSL data";
+        break;
+    case CURLINFO_DATA_IN:
+        text = "<= Recv data";
+        break;
+    case CURLINFO_SSL_DATA_IN:
+        text = "<= Recv SSL data";
+        break;
+    }
+    TRACE_VERBOSE("%s:%d %s\n", __func__, __LINE__, text);
+
+    return 0;
+}
+
+WINHTTPAPI HINTERNET WINAPI WinHttpOpen
+(
+    LPCTSTR pszAgentW,
+    DWORD dwAccessType,
+    LPCTSTR pszProxyW,
+    LPCTSTR pszProxyBypassW,
+    DWORD dwFlags
+)
+{
+    std::shared_ptr<WinHttpSessionImp> session = std::make_shared<WinHttpSessionImp> ();
+
+    TRACE("%s:%d\n", __func__, __LINE__);
+    if (dwAccessType == WINHTTP_ACCESS_TYPE_NAMED_PROXY)
+    {
+        if (!pszProxyW)
+            return NULL;
+
+        ConvertCstrAssign(pszProxyW, WCTLEN(pszProxyW), session->GetProxy());
+
+        std::vector<std::string> proxies = Split(session->GetProxy(), ';');
+        session->SetProxies(proxies);
+    }
+
+    if (dwFlags & WINHTTP_FLAG_ASYNC) {
+        session->SetAsync();
+    }
+
+    WinHttpHandleContainer<WinHttpSessionImp>::Instance().Register(session);
+    return session.get();
+}
+
+WINHTTPAPI HINTERNET WINAPI WinHttpConnect
+(
+    HINTERNET hSession,
+    LPCTSTR pswzServerName,
+    INTERNET_PORT nServerPort,
+    DWORD dwReserved
+)
+{
+    WinHttpSessionImp *session = static_cast<WinHttpSessionImp *>(hSession);
+
+    TRACE("%s:%d pswzServerName: " STRING_LITERAL " nServerPort:%d\n", __func__, __LINE__, pswzServerName, nServerPort);
+    ConvertCstrAssign(pswzServerName, WCTLEN(pswzServerName), session->GetServerName());
+
+    session->SetServerPort(nServerPort);
+    std::shared_ptr<WinHttpConnectImp> connect = std::make_shared<WinHttpConnectImp> ();
+    if (!connect)
+        return NULL;
+
+    connect->SetHandle(session);
+    WinHttpHandleContainer<WinHttpConnectImp>::Instance().Register(connect);
+    return connect.get();
+}
+
+
+BOOLAPI WinHttpCloseHandle(
+    HINTERNET hInternet
+)
+{
+    WinHttpBase *base = static_cast<WinHttpBase *>(hInternet);
+
+    TRACE("%s:%d\n", __func__, __LINE__);
+    if (!base)
+        return FALSE;
+
+    if (WinHttpHandleContainer<WinHttpRequestImp>::Instance().IsRegistered(reinterpret_cast<WinHttpRequestImp *>(hInternet)))
+    {
+        WinHttpRequestImp *request = dynamic_cast<WinHttpRequestImp *>(base);
+        if (!request)
+            return FALSE;
+
+        std::shared_ptr<WinHttpRequestImp> srequest = request->shared_from_this();
+        if (!srequest)
+            return FALSE;
+
+        request->GetClosing() = true;
+        WinHttpHandleContainer<WinHttpRequestImp>::Instance().UnRegister(request);
+        return TRUE;
+    }
+
+    if (WinHttpHandleContainer<WinHttpSessionImp>::I
```

Comment:
Might as well combine declaration with initialization -- occurs many times in this file.